### DISCUSSION TITLE
I'm wondering how the order of arguments is determined of a SAM when using the Adapter annotation. Feels like a bug, but not sure.

### DISCUSSION THREAD

**Discussion by willemsst** [micronaut-projects/micronaut-core#4254] [Q&A] [answered]
I've been fighting this morning with the use of the `@Adapter` annotation and using a `Qualifiers.byTypeArguments(..)` to lookup my generated _Single Abstract Method_ **(SAM)** classes.

For some weird reason, the order of types in the generated types, is not the same as the one in my interface that is referenced by the @Adapter annotation.

I've been trying to find the magic of the AOP code generation in the source code, but couldn't spot it yet, so, if anyone can clarify how the order is determined, I would like to look into it to understand what's going on. Because, it feels like a bug that the order is changed, but as this is already at version 2.1.0, I might just do something stupid as well.

**Comment by graemerocher** [comment, marked as answer]
The logic for this can be found here:

https://github.com/micronaut-projects/micronaut-core/blob/1c2f8783cb00da30890330aaf8fd267a1cf208b4/inject-java/src/main/java/io/micronaut/annotation/processing/BeanDefinitionInjectProcessor.java#L1410

You can debug it running these tests:

https://github.com/micronaut-projects/micronaut-core/blob/1c2f8783cb00da30890330aaf8fd267a1cf208b4/inject-java/src/test/groovy/io/micronaut/aop/adapter/MethodAdapterSpec.groovy#L27

**Reply by willemsst** [reply]
Thank you for the quick response.

I've been adding my own classes to the test suite, added a test with multiple arguments in the event listener interface and debugged it. It was quite a deep dive, haha. But I found and resolved the issue.

It's a bug.

https://github.com/micronaut-projects/micronaut-core/blob/1c2f8783cb00da30890330aaf8fd267a1cf208b4/inject-java/src/main/java/io/micronaut/annotation/processing/BeanDefinitionInjectProcessor.java#L1474

-> this should be something like a **`LinkedHashMap`** in my opinion so the order of arguments is respected.

Because, in the next block of code, the `HashMap` with arguments is used to generate the `getTypeArguments()` method on the generated proxy class.

https://github.com/micronaut-projects/micronaut-core/blob/1c2f8783cb00da30890330aaf8fd267a1cf208b4/inject-java/src/main/java/io/micronaut/annotation/processing/BeanDefinitionInjectProcessor.java#L1527-L1535

**Reply by graemerocher** [reply]
Great diagnosis! Would you like to send a PR?

**Reply by willemsst** [reply]
done. (i hope I did it correct as it's my first ever)
build check failed on Kotlin/Http stuff 🤷‍♂️
license/ca signed with multiple attempts but it's still in pending 🤷‍♂️ 

I hope you can move on with it. I won't put more time into this.
